File: dart_client/lib/src/services/auth_service.dart
import 'package:dio/dio.dart';
import '../dtos.dart';
import '../enums.dart';

class AuthService {
  final Dio _dio;
  AuthService(this._dio);

  Future<DTOResponse<LoginResponseDTO>> postLogin({required LoginRequestDTO body}) async {
    final response = await _dio.post('/auth/login', data: body.toJson());
    return DTOResponse<LoginResponseDTO>.fromJson(response.data, (json) => LoginResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<LoginResponseDTO>> postSignup({required SignupRequestDTO body}) async {
    final response = await _dio.post('/auth/signup', data: body.toJson());
    return DTOResponse<LoginResponseDTO>.fromJson(response.data, (json) => LoginResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<GenericResponseDTO>> postLogout({}) async {
    final response = await _dio.post('/auth/logout');
    return DTOResponse<GenericResponseDTO>.fromJson(response.data, (json) => GenericResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<MessageResponseDTO>> deleteAccount({}) async {
    final response = await _dio.delete('/auth/account');
    return DTOResponse<MessageResponseDTO>.fromJson(response.data, (json) => MessageResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<MessageResponseDTO>> postRecover({required RecoverPasswordRequestDTO body}) async {
    final response = await _dio.post('/auth/recover', data: body.toJson());
    return DTOResponse<MessageResponseDTO>.fromJson(response.data, (json) => MessageResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<GenericResponseDTO>> postVerify({required VerifyEmailRequestDTO body}) async {
    final response = await _dio.post('/auth/verify', data: body.toJson());
    return DTOResponse<GenericResponseDTO>.fromJson(response.data, (json) => GenericResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<MustChangePasswordResponseDTO>> getMustChangePassword({}) async {
    final response = await _dio.get('/auth/must-change-password');
    return DTOResponse<MustChangePasswordResponseDTO>.fromJson(response.data, (json) => MustChangePasswordResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<GenericResponseDTO>> postChangePassword({required ChangePasswordRequestDTO body}) async {
    final response = await _dio.post('/auth/change-password', data: body.toJson());
    return DTOResponse<GenericResponseDTO>.fromJson(response.data, (json) => GenericResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<UserStatusResponseDTO>> getStatus({}) async {
    final response = await _dio.get('/auth/status');
    return DTOResponse<UserStatusResponseDTO>.fromJson(response.data, (json) => UserStatusResponseDTO.fromJson(json as Map<String, dynamic>));
  }

}

File: js_client/contracts/services/auth_service.js
import { DTOResponse, DTOParams, EmptyParamsDTO, SearchQuestionsParamsDTO, GetQuestionParamsDTO, GetQuestionsBatchParamsDTO, UpdateQuestionParamsDTO, DeleteQuestionParamsDTO, UpdateCollectionParamsDTO, EmptyRequestDTO, BaseResponse, LoginResponseDTO, GenericResponseDTO, MessageResponseDTO, MustChangePasswordResponseDTO, UserStatusResponseDTO, ProfileDataResponseDTO, QuestionListResponseDTO, QuestionDataResponseDTO, CollectionListResponseDTO, IdDataResponseDTO, CollectionDetailResponseDTO, MarkListResponseDTO, FriendRequestResponseDTO, UserListResponseDTO, LoginRequestDTO, LoginDataDTO, SignupRequestDTO, MessageDataDTO, RecoverPasswordRequestDTO, VerifyEmailRequestDTO, MustChangePasswordDataDTO, ChangePasswordRequestDTO, UserStatusDataDTO, RegisterPushTokenRequestDTO, ProfileDataDTO, UpdateProfileRequestDTO, AnswerDataDTO, QuestionDataDTO, LocalizationDTO, CreateAnswerInputDTO, CreateQuestionInputDTO, CreateQuestionsRequestDTO, UpdateAnswerInputDTO, UpdateQuestionRequestDTO, CollectionDataDTO, CollectionDetailDataDTO, CreateCollectionRequestDTO, IdDataDTO, UpdateCollectionRequestDTO, MarkDataDTO, SendFriendRequestDTO, FriendRequestDataDTO, RespondFriendRequestRequestDTO, FriendRequestListResponseDTO, PublicUserProfileDTO, ErrorDetailsDTO } from '../dtos.js';
import * as Enums from '../enums.js';

export class AuthService {
  constructor(baseUrl, fetcher) {
    this.baseUrl = baseUrl;
    this.fetcher = fetcher;
  }

  async postLogin({ body,  } = {}) {
    const url = new URL(`${this.baseUrl}/auth/login`);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'POST',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => LoginResponseDTO.fromJson(data));
  }

  async postSignup({ body,  } = {}) {
    const url = new URL(`${this.baseUrl}/auth/signup`);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'POST',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => LoginResponseDTO.fromJson(data));
  }

  async postLogout({ bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/auth/logout`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'POST',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => GenericResponseDTO.fromJson(data));
  }

  async deleteAccount({ bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/auth/account`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'DELETE',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => MessageResponseDTO.fromJson(data));
  }

  async postRecover({ body,  } = {}) {
    const url = new URL(`${this.baseUrl}/auth/recover`);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'POST',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => MessageResponseDTO.fromJson(data));
  }

  async postVerify({ body,  } = {}) {
    const url = new URL(`${this.baseUrl}/auth/verify`);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'POST',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => GenericResponseDTO.fromJson(data));
  }

  async getMustChangePassword({ bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/auth/must-change-password`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'GET',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => MustChangePasswordResponseDTO.fromJson(data));
  }

  async postChangePassword({ body, bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/auth/change-password`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'POST',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => GenericResponseDTO.fromJson(data));
  }

  async getStatus({ bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/auth/status`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'GET',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => UserStatusResponseDTO.fromJson(data));
  }

}
